File: sdk/go/harness/gemini.go
package harness

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// GeminiProvider invokes the Gemini CLI as a subprocess.
// Gemini outputs plain text (no structured JSON), so results are single-turn.
type GeminiProvider struct {
	BinPath string
}

// NewGeminiProvider creates a Gemini provider. If binPath is empty,
// it defaults to "gemini".
func NewGeminiProvider(binPath string) *GeminiProvider {
	if binPath == "" {
		binPath = "gemini"
	}
	return &GeminiProvider{BinPath: binPath}
}

func (p *GeminiProvider) Execute(ctx context.Context, prompt string, options Options) (*RawResult, error) {
	cmd := []string{p.BinPath}

	if options.Cwd != "" {
		cmd = append(cmd, "-C", options.Cwd)
	} else if options.ProjectDir != "" {
		cmd = append(cmd, "-C", options.ProjectDir)
	}

	if options.PermissionMode == "auto" {
		cmd = append(cmd, "--sandbox")
	}

	// gemini has no reasoning-effort flag; strip any "#variant" suffix so
	// the CLI still receives a valid model id, and drop the variant.
	modelValue, _ := options.resolveModelAndVariant()
	if modelValue != "" {
		cmd = append(cmd, "-m", modelValue)
	}

	// Prompt passed via -p flag.
	cmd = append(cmd, "-p", prompt)

	env := make(map[string]string)
	for k, v := range options.Env {
		env[k] = v
	}

	cwd := options.Cwd
	if cwd == "" {
		cwd = options.ProjectDir
	}

	startAPI := time.Now()

	cliResult, err := RunCLI(ctx, cmd, env, cwd, options.timeout())
	apiMS := int(time.Since(startAPI).Milliseconds())

	if err != nil {
		if isExecNotFound(err) {
			return &RawResult{
				IsError: true,
				ErrorMessage: fmt.Sprintf(
					"Gemini binary not found at '%s'. Install: https://github.com/google-gemini/gemini-cli",
					p.BinPath,
				),
				FailureType: FailureCrash,
				Metrics:     Metrics{},
			}, nil
		}
		if strings.Contains(err.Error(), "timed out") {
			return &RawResult{
				IsError:      true,
				ErrorMessage: err.Error(),
				FailureType:  FailureTimeout,
				Metrics:      Metrics{DurationAPIMS: apiMS},
			}, nil
		}
		return nil, err
	}

	resultText := strings.TrimSpace(cliResult.Stdout)
	cleanStderr := StripANSI(strings.TrimSpace(cliResult.Stderr))

	raw := &RawResult{
		Result:   resultText,
		Messages: nil,
		Metrics: Metrics{
			DurationAPIMS: apiMS,
			NumTurns:      1,
		},
		ReturnCode: cliResult.ReturnCode,
	}

	if cliResult.ReturnCode < 0 {
		raw.FailureType = FailureCrash
		raw.IsError = true
		if cleanStderr != "" {
			raw.ErrorMessage = fmt.Sprintf("Process killed by signal %d. stderr: %.500s",
				-cliResult.ReturnCode, cleanStderr)
		} else {
			raw.ErrorMessage = fmt.Sprintf("Process killed by signal %d.", -cliResult.ReturnCode)
		}
	} else if cliResult.ReturnCode != 0 && resultText == "" {
		raw.FailureType = FailureCrash
		raw.IsError = true
		if cleanStderr != "" {
			raw.ErrorMessage = truncate(cleanStderr, 1000)
		} else {
			raw.ErrorMessage = fmt.Sprintf("Process exited with code %d and produced no output.", cliResult.ReturnCode)
		}
	}

	if resultText == "" {
		raw.Metrics.NumTurns = 0
	}

	return raw, nil
}

File: sdk/typescript/src/harness/providers/gemini.ts
import type { HarnessProvider } from './base.js';
import type { RawResult } from '../types.js';
import { createRawResult, createMetrics } from '../types.js';
import { runCli } from '../cli.js';
import { resolveModelAndVariant } from '../modelVariant.js';

export class GeminiProvider implements HarnessProvider {
  private readonly bin: string;

  constructor(binPath = 'gemini') {
    this.bin = binPath;
  }

  async execute(prompt: string, options: Record<string, unknown>): Promise<RawResult> {
    const cmd = [this.bin];

    if (options.cwd) {
      cmd.push('-C', String(options.cwd));
    }
    if (options.permissionMode === 'auto') {
      cmd.push('--sandbox');
    }
    // gemini has no reasoning-effort flag; strip any "#variant" suffix so
    // the CLI still receives a valid model id.
    const { model: modelValue } = resolveModelAndVariant(options);
    if (modelValue) {
      cmd.push('-m', modelValue);
    }
    cmd.push('-p', prompt);

    const startApi = Date.now();
    try {
      const { stdout, stderr, exitCode } = await runCli(cmd, {
        env: options.env as Record<string, string> | undefined,
        cwd: options.cwd as string | undefined,
      });

      const resultText = stdout.trim() || undefined;
      const isError = exitCode !== 0 && !resultText;

      return createRawResult({
        result: resultText,
        messages: [],
        metrics: createMetrics({
          durationApiMs: Date.now() - startApi,
          numTurns: resultText ? 1 : 0,
          sessionId: '',
        }),
        isError,
        errorMessage: isError ? stderr.trim() : undefined,
      });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('ENOENT')) {
        return createRawResult({
          isError: true,
          errorMessage: `Gemini binary not found at '${this.bin}'. Install: https://github.com/google-gemini/gemini-cli`,
          metrics: createMetrics({ durationApiMs: Date.now() - startApi }),
        });
      }
      return createRawResult({
        isError: true,
        errorMessage: msg,
        metrics: createMetrics({ durationApiMs: Date.now() - startApi }),
      });
    }
  }
}
